crashtracker/src/receiver.rs

```diff
@@ -283,30 +296,64 @@ enum CrashReportStatus {
 /// (for instance if it crashed while calculating the crash-report), we return
 /// a PartialCrashReport.
 async fn receive_report(
+    timeout: Duration,
     stream: impl AsyncBufReadExt + std::marker::Unpin,
 ) -> anyhow::Result<CrashReportStatus> {
     let mut crashinfo = CrashInfo::new();
     let mut stdin_state = StdinState::Waiting;
     let mut config = None;
 
     let mut lines = stream.lines();
+    let mut deadline = None;
+    // Start the timeout counter when the deadline when the first crash message is recieved
+    let mut remaining_timeout = Duration::MAX;
 
     //TODO: This assumes that the input is valid UTF-8.
-    while let Some(line) = lines.next_line().await? {
+    loop {
+        let next = tokio::time::timeout(remaining_timeout, lines.next_line()).await;
+        if let Err(elapsed) = next {
+            eprintln!("Timeout: {elapsed}");
+            break;
+        };
+        let next = next.unwrap();
+        if let Err(io_err) = next {
+            eprintln!("IO Error: {io_err}");
+            break;
+        }
+        let next = next.unwrap();
+        if next.is_none() {
+            break;
+        }
+        let line = next.unwrap();
+
         match process_line(&mut crashinfo, &mut config, line, stdin_state) {
-            Ok(next_state) => stdin_state = next_state,
+            Ok(next_state) => {
+                stdin_state = next_state;
+                if matches!(stdin_state, StdinState::Done) {
+                    break;
+                }
+            }
             Err(e) => {
                 // If the input is corrupted, stop and salvage what we can
                 stdin_state = StdinState::InternalError(e.to_string());
                 break;
             }
         }
+        if let Some(deadline) = deadline {
+            // The clock was already ticking, update the remaining time
+            remaining_timeout = deadline - Instant::now()
+        } else {
+            // We've recieved the first message from the collector, start the clock ticking.
+            deadline = Some(Instant::now() + timeout);
+            remaining_timeout = timeout;
+        }
     }
 
     if !crashinfo.crash_seen() {
         return Ok(CrashReportStatus::NoCrash);
     }
 
+    // Without a config, we don't even know the endpoint to transmit to.  Not much to do to recover.
```

Comment:
We _could_ encode this in the args or the environment, as per my earlier comment about the prepared args.

That said, I think there's not much to do in this case anyway and the likelihood is low anyway.